### ISSUE TITLE
Add brew packages to laptop

### ISSUE THREAD

**Issue by petelindgren** [Mariana-Tek/laptop#7]
Recommend making sure homebrew installs (for **Big Sur**)
- [x] pre-commit
- [x] pyenv
- [x] pipenv
These are not installed by the laptop script if version is **Big Sur**

these are added by Jeremy's custom script.
recommend adding this code block after git section

```sh
## Python Virtual Environments and pre-commit
cask_install_or_skip 'pyenv' 'pyenv'
cask_install_or_skip 'pipenv' 'pipenv'
cask_install_or_skip 'pre-commit' 'pre-commit'
```

### Update README.rst
- [ ] Add comment about `xcode-select --install` not being necessary if on Mac bigSur.  You get a message that everything is already installed.
- [ ] maybe add a comment about running this if `psycopg2` does not install with pipenv:
  ```sh
  brew link openssl
  echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
  export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
  export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
  export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig"
  ```

**Comment by bashhack** [comment]
@petelindgren - Thanks for providing some feedback! I've tried to strike a middle-ground and work toward the spirit of your requests! :pray: 

### pre-commit
`pre-commit` so far is really a dependency only of Mariana Django and is already included when installing that project (see dev dependencies in Pipfile) - so for now, I'd like to avoid bringing that into the `laptop` script. Can you think of another reason to include here as a recommended default?

### pycopg2
Similar thoughts for me apply to the request to add info about `psycopg2` - this script is intended to generally be applicable for an engineer to get up and running with the baseline machine setup as quickly as possible so that they can then focus on project-specific installations shortly thereafter. I'm hesitant to include very specific tips on installation of dependencies that may not be used across all projects.

### xcode
I think based on personal experience with new devs and new machines in the last 4 - 6 months, it is reasonable and easy insurance to retain the `xcode-select --install` verbiage. For that percentage of time in which it is strictly required on a newly shipped Mac, not having it documented seems to create an extra hurdle and time investment for myself or others to debug and ultimately suggest they run that command.

I also hear your perspective, and I want to honor it - I've tried to strike a balance by updating the README with the following to add clarity by adding a NOTE about the `xcode-select` steps with console output, see:

https://github.com/Mariana-Tek/laptop#requirements

### Python toolchain
Regarding the Python toolchain, the reality right now based on upstream vendor issues is that including the Python ecosystem dependencies is just far trickier than we'd hope for. There's a host of issues around both pyenv and pipenv on M1 Macs, for example and Big Sur on Intel with Python versions less than 3.8 are totally borked without additional effort. Before adding the information you'll see below, I wiped my laptop and did a fresh install of Big Sur and this laptop script to confirm for accuracy.

With that said, I feel pretty strongly that we should continue skipping the automatic installation for Big Sur users, unfortunately. However, I hope you'll find that I've attempted to meet you halfway by adding additional information for that subset of users:

[Big Sur (Intel) Addendum](https://github.com/Mariana-Tek/laptop#big-sur-intel-addendum)

**Comment by petelindgren** [comment]
@bashhack - I am provisioning a new Mariana-Tek laptop right now because my current one has problems.

Update your .zshrc file and append this:
```sh
# Export openssl to help build pipenv
export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"
# Export postgresql@11 to path so pipenv can create virtual environments using pg_config to compile psycopg2
export PATH="/usr/local/opt/postgresql@11/bin:$PATH"

export PATH="$HOME/.pyenv/bin:$PATH"
export PATH="/usr/local/bin:$PATH"

# Export 
export LDFLAGS="-L/usr/local/opt/zlib/lib -L/usr/local/opt/bzip2/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/postgresql@11/lib"
export CPPFLAGS="-I/usr/local/opt/zlib/include -I/usr/local/opt/bzip2/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/postgresql@11/include"

export PKG_CONFIG_PATH="/usr/local/opt/postgresql@11/lib/pkgconfig"

# (The below instructions are intended for common
# shell setups. See the README for more guidance
# if they don't apply and/or don't work for you.)

# Add pyenv executable to PATH and
# enable shims by adding the following
# to ~/.profile and ~/.zprofile:

export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"

# Load pyenv into the shell by adding
# the following to ~/.zshrc:

eval "$(pyenv init -)"
```


your latest updates did not let me `pyenv install 3.6.8`, but it looks real close.  Not sure if the problem was the smart quotes or I needed to reinstall more

This link https://gist.github.com/aputs/1e898436c70c23ea739899a7942a720a had the secret to `pyenv install 3.6.8` though I made one change which was from `brew install ...` to `brew reinstall ...`


```
export MACOSX_DEPLOYMENT_TARGET=11.0
brew reinstall pyenv pipenv bzip2 zlib xz openssl@1.1 postgresql@11 pre-commit

## 3.6.8
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.6.8 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)

## 3.6.12
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.6.12 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)

## 3.7.9
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install 3.7.9
```

**Comment by petelindgren** [comment]
@bashhack , here are some proposed updates
- [x] simplify brew reinstall packages to include `pre-commit` and a few other packages like `postgresql@11`
- [x] Fix smart quotes that do not install 3.6.8 in some shell environments
- [x] Add section for updating `.zshrc` file

since I do not have PR permissions to this public repo (probably smart) I recommend updating the Big Sur Addendum in the README.md file to this:

Big Sur (Intel) Addendum
------------------------

Since late 2020, with the launch of Apple's Big Sur OS release, there are a number of machine-specific errors one may encounter in relation to Python dependencies that are documented widely in upstream vendor bug tickets and across Github.

As we can't predict either when or if these will be resolved, the following is a rough guide to the post-`laptop` script steps we currently (as of May 2021) recommend:

```sh
# Install/Reinstall Python tools
brew reinstall pre-commit pyenv pipenv bzip2 zlib xz openssl@1.1 postgresql@11 chamber

# Install Python 3.6.8
# NOTE: 
# This is where the bulk of the issues arise - if you are installing most Python versions > 3.8, you may find that a simple `pyenv install 3.8` or `pyenv install 3.9` is sufficient
# However, for versions < 3.8, additional work is needed for the OS to compile Python properly. Shown below is the command needed as of May 2021 for Python 3.6.8 (our primary Python
# version for the Mariana Django project)
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.6.8 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
```

Highly recommend installing **Oh My Zsh** https://ohmyz.sh/#install to replace the default `.zsh`

Append this to the end of the `.zshrc` file in your home directory (even if you do not install **Oh My Zsh** https://ohmyz.sh/#install)
```sh
# Export openssl to help build pipenv
export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"
# Export postgresql@11 to path so pipenv can create virtual environments using pg_config to compile psycopg2
export PATH="/usr/local/opt/postgresql@11/bin:$PATH"

# For nvm (required for ember app)
export NVM_DIR="$HOME/.nvm"
[ -s "/usr/local/opt/nvm/nvm.sh" ] && . "/usr/local/opt/nvm/nvm.sh"  # This loads nvm
[ -s "/usr/local/opt/nvm/etc/bash_completion.d/nvm" ] && . "/usr/local/opt/nvm/etc/bash_completion.d/nvm"  # This loads nvm bash_completion
```

**Comment by bashhack** [comment]
Since M1 setup is Docker-based, closing for now as the Python environment configurationrecommended no longer requires any local install via pyenv, pipenv, etc.
